lib/assets/Asset.js

```diff
@@ -114,6 +114,39 @@ class Asset extends EventEmitter {
         }
     }
 
+    get type() {
+        if (this._type === 'Asset') {
+            if (this.hasOwnProperty('contentType')) {
+                const typeFromContentType = AssetGraph.lookupContentType(this.contentType);
+                if (typeFromContentType && typeFromContentType !== 'Asset') {
+                    return typeFromContentType;
+                }
+            }
+            let commonTargetType;
+            for (const incomingRelation of this.incomingRelations) {
+                if (incomingRelation.targetType) {
+                    if (!commonTargetType) {
+                        commonTargetType = incomingRelation.targetType;
+                    } else if (commonTargetType !== incomingRelation.targetType) {
+                        commonTargetType = undefined;
+                        break;
+                    }
+                }
+            }
+            const typeFromExtension = this.url && AssetGraph.typeByExtension[pathModule.extname(this.url.replace(/[\?\#].*$/, '')).toLowerCase()];
+            if (typeFromExtension && (!commonTargetType || AssetGraph[commonTargetType].prototype['is' + commonTargetType])) {
+                return typeFromExtension;
+            } else if (commonTargetType) {
+                return commonTargetType;
+            }
+        }
+        return this._type;
+    }
+
+    set type(type) {
+        this._type = type;
```

Comment:
Should there be a setter at all, and if so, should this not include the logic to detect sub typing or switches to different asset type family branches?

Reply:
I assumed that there would be some places where the `type` property would be set by `createAsset` or friends, but apparently I didn't bother to check. It works fine without the setter, and I've removed it now.